### PR TITLE
Allows running under another user and non-raspberrypi hardware

### LINKED ISSUE #108: Installing for Non "pi" User
After running the install_and_upgrade.sh script I get the following error:

Processing triggers for libc-bin (2.28-10+rpi1) ...
 ✓   Ansible install complete!
 * Checking for configuration settings...
 ✓   Settings file ready!
 * Running Ansible to install and/or update your raspberry-noaa-v2...

PLAY [install raspberry-noaa-v2 core] ********************************************************************************************************************************************************************************************************

TASK [get os release] ************************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [common : noaa environment variables] ***************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {"changed": false, "checksum": "fbab5a3e66f5ff73e9027e71766d9376ea5724f7", "msg": "Destination directory ~pi does not exist"}
	to retry, use: --limit @/home/not-pi/raspberry-noaa-v2/ansible/site.retry

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=1    changed=0    unreachable=0    failed=1   

Error:   Something failed with the install - please inspect the logs above

How do I get around this?

### PR DESCRIPTION
**Pull request by rhaamo**
I wanted to install this on a server I already had and not a dedicated pi so I tried to make the whole install part and run more user/hw/sw agnostic.

TLDR: made able to run on another user than `pi` and hardware than a pi.

My changes are:
- enabled facts gathering (I used some)
- added own build of medet for amd64 (you can replace it if you want)
- added deb for wxtoimg for amd64 (same)
- install of the last two depending on if the CPU arch from the facts is `armv7l` or not (I only have one pi arch to test the value content)
- added a few config/ansible variables:
  - `running_under_user`, default `pi`, the user the app will run under
  - `running_under_group`, same but for the group
  - `use_ntpd`: default `true`, will include or not ntp.yml
  - `use_ramfs`: default `true`, will include or not ramfs.yml
  - `setup_webserver`: default `true`, include nginx/php install and setup
- nginx & php install/setup has been split from the app config
- replacement of `/home/pi` in two files by ansible using the `running_under_user` var
- used `running_under_user` instead of hardcoded `pi` in various ansible templates
- added a check for another path for `rtl_sdr` (was on /usr/bin instead of /usr/local/bin here)
- replaced `raspbian_version` by fact `ansible_distribution_release`
- predict is installed from repos when stretch, and from stretch deb file if != stretch (not packaged anymore after stretch)
- added the user and group running under in `noaa-v2.conf` to be used for install script and anything else that could use it
- added the user in `Config/Config.php` and used it in `AdminController.php` for sudo command

NOTES:

- One of the issues I see would be with the subst in two files (see `common/tasks/user.yml`) that will break `git pull` but I don't see how to properly handle that, except by excluding the scripts from git (the `scripts/xxxx.ext`), making them templates (`scripts/xxxx.ext.j2`), and having ansible generating the 'final' ones.
- Also in `scripts/common.sh`, `rtl_fm` isn't always in `/usr/local/bin` (like `rtl_sdr`) but might be in `/usr/bin`, maybe we can add detection in ansible and fill `noaa-v2.conf` with it.

- Documentation has not been updated so depending on the config used (`setup_webserver` etc.) the install will not be 100% operational.

- This PR should also fixes #108

- I don't have an available raspberrypi to test it, it only has been tested on my server (running Ubuntu 20.04.3 LTS)
- Gnuradio scripts are for python2 BUT seems easy to use python3, noaa one needs indentation fixed (mix tabs/spaces) + python3 on first line, and only first line for meteor

Here is one okay-ish (the pass seemed to have been shorted because of an overlapping meteor one) pass from my install:
image

### CODE REVIEW

**Comment by dom-robinson** [comment]
Wow - what an awesome PR to come in out of the blue! Thank you so much!

This is quite an involved development to pick through and test, and we are all working on this in spare time so it may take some time to build the momentum to get this through to live, but strongly encourage it to happen!

@colinluthier @Cadair check this out!

Dom

**Comment by Cadair** [comment]
:heart: Thank you so much for this, it looks great. I shall try and look over it (and maybe even try and deploy it on my server) this weekend and see how I go.

File: ansible/group_vars/all.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@
 available_langs: [ ar, de, en, es, sr ]
 
 # root git clone directory (where functionality exists)
-noaa_home: "/home/pi/raspberry-noaa-v2"
+noaa_home: "/home/{{ running_under_user }}/raspberry-noaa-v2"
```
Comment (Cadair) [inline ansible/group_vars/all.yml]: We are still assuming that the playbook is executed by the user we are installing as correct?

In that case I think this would be a better solution:
```suggestion
noaa_home: "{{ lookup('env', 'HOME') }}/raspberry-noaa-v2"
```

File: config/settings.yml.sample [hunk 2]
```diff
@@ -1,4 +1,8 @@
 ---
+# User under the app will run, default is 'pi' for a raspberry
+running_under_user: pi
+running_under_group: pi
```
Comment (Cadair) [inline config/settings.yml.sample]: Why don't we default these to the user that executes ansible?

File: ansible/roles/common/tasks/user.yml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+---
```
Comment (Cadair) [inline ansible/roles/common/tasks/user.yml:1]: I think we should probably change these scripts so that they use the home directory env var in the context they are executed and not add this here?

File: ansible/roles/common/tasks/tools.yml [hunk 4]
```diff
@@ -1,10 +1,19 @@
 ---
-- name: install wxtoimg
+- name: install wxtoimg arm
   become: yes
   apt:
     deb: "{{ noaa_home }}/software/wxtoimg-armhf-2.11.2-beta.deb"
     state: present
     update_cache: no
+  when: ansible_architecture == 'armv7l'
```
Comment (Cadair) [inline ansible/roles/common/tasks/tools.yml:8]: huh, so we didn't support a 64bit OS for the pi 4 then?

File: ansible/roles/common/tasks/dependencies.yml [hunk 5]
```diff
@@ -36,20 +35,48 @@
...
+  when: ansible_architecture == 'armv7l'
+
+- name: install wkhtmltox (wkhtmltoimg) x86_64
+  become: yes
+  apt:
+    deb: "https://github.com/wkhtmltopdf/packaging/releases/download/0.12.6-1/wkhtmltox_0.12.6-1.{{ ansible_distribution_release }}_amd64.deb"
+    state: present
+  when: ansible_architecture == 'x86_64'
+
+# predict is not packaged anymore after stretch
```
Comment (Cadair) [inline ansible/roles/common/tasks/dependencies.yml:65]: :frowning_face:

**Review by Cadair — COMMENTED** [review]
Some initial comments without looking too deeply.

File: ansible/roles/common/tasks/dependencies.yml (hunk 5 above)
Comment (rhaamo) [reply ansible/roles/common/tasks/dependencies.yml:65]: for predict ? the debian package seems to have a bunch of patches to it and is even older than latest from website. I don't really know what would be the best plan, maybe using pypredict in a script

File: ansible/roles/common/tasks/tools.yml (hunk 4 above)
Comment (rhaamo) [reply ansible/roles/common/tasks/tools.yml:8]: I've got the value from one of the pi I have that is probably an old v3 something, but I think armhf should run on a 64bit pi 4

File: ansible/roles/common/tasks/dependencies.yml (hunk 5 above)
Comment (Cadair) [reply ansible/roles/common/tasks/dependencies.yml:65]: Yeah, it's a shame the debian package became unmaintained.

What you have done here is probably fine in the short term, we should perhaps open a new issue discussing other options though?

File: ansible/roles/common/tasks/user.yml (hunk 3 above)
Comment (rhaamo) [reply ansible/roles/common/tasks/user.yml:1]: if they accepts relative paths we can use `~/` instead of full one yes, I could try it in the grc (I don't use the twitter part) if it's called on a standard pass

File: config/settings.yml.sample (hunk 2 above)
Comment (rhaamo) [reply config/settings.yml.sample]: same as the other comment, I guess we can default them if we assume running under the final user, that would simplify a bit the configuration

**Comment by Cadair** [comment]
> Gnuradio scripts are for python2 BUT seems easy to use python3, noaa one needs indentation fixed (mix tabs/spaces) + python3 on first line, and only first line for meteor

OH that's why that worked!! Stupid Python 2 still existing.

File: ansible/roles/common/tasks/dependencies.yml (hunk 5 above)
Comment (rhaamo) [reply ansible/roles/common/tasks/dependencies.yml:65]: absolutely, in my hackish pile of python it did use pypredict, I haven't looked here how and when is predict used, but since there is already python it could be added with the rest.

Comment (Cadair) [reply ansible/roles/common/tasks/dependencies.yml:65]: :+1: Let's move this here: https://github.com/jekhokie/raspberry-noaa-v2/issues/462 it should happen irrespective of this PR :smile_cat:

File: ansible/roles/common/tasks/user.yml (hunk 3 above)
Comment (Cadair) [reply ansible/roles/common/tasks/user.yml:1]: Or we can use one of the Python ways of getting the home dir e.g. `os.path.expanduser("~")`. I think I would prefer trying to do that in this PR rather than this trick :see_no_evil:

File: config/settings.yml.sample (hunk 2 above)
Comment (Cadair) [reply config/settings.yml.sample]: Well I think we are assuming that people are doing that at the moment with the pi user. So I think it's probably a good default to use, and then people who want to do more exotic things can set the var or invoke ansible themselves etc?

File: ansible/group_vars/all.yml (hunk 1 above)
Comment (Cadair) [reply ansible/group_vars/all.yml]: The alternative if we don't want to make that assumption is to do something fruity like this: https://stackoverflow.com/a/45447488 with getent and set_fact?

File: config/settings.yml.sample (hunk 2 above)
Comment (rhaamo) [reply config/settings.yml.sample]: thinking a bit more about that (and the two other comments) I'm not sure it would work with another user than the one it's ran under, mostly because the scripts aren't moving and are "expected" to be under the current user.

I think we can move thoses two out of settings.yml, and use two facts on the group vars defaults file:
```
# from an `ansible -m setup localhost`
        "ansible_user_dir": "/home/dashie",
        "ansible_user_id": "dashie",
```

Comment (Cadair) [reply config/settings.yml.sample]: Sounds good, I would kinda like to make this run as a proper ansible role one day, but we can cross that bridge when we come to it :+1:

Comment (rhaamo) [reply config/settings.yml.sample]: great, I will change that part a bit later today

File: ansible/roles/common/tasks/user.yml (hunk 3 above)
Comment (rhaamo) [reply ansible/roles/common/tasks/user.yml:1]: done for the python script, I'm testing with `~/` in the grc and will see on the next meteor pass if I see any error in the logs

File: scripts/tools/sun.py [hunk 6]
```diff
@@ -7,7 +7,7 @@
 from envbash import load_envbash
 
 # load bash environment vars
-load_envbash('/home/pi/.noaa-v2.conf')
+load_envbash("%s/.noaa-v2.conf" % os.path.expanduser('~'))
```
Comment (Cadair) [inline scripts/tools/sun.py]:
```suggestion
load_envbash("{}/.noaa-v2.conf".format(os.path.expanduser('~')))
```

**Comment by rhaamo** [comment]
I have fixed `ramfs_path` missing the `{{ }}` and used the vars for user and group.

The METEOR stuff uses the ramfs path, so if not choosen to be deployed the user handle it, and if choosen it will have the right user and group instead of fixed `pi` (somehow missed that)

File: ansible/roles/common/tasks/tools.yml (hunk 4 above)
Comment (Cadair) [reply ansible/roles/common/tasks/tools.yml:8]: Apparently not? https://github.com/jekhokie/raspberry-noaa-v2/issues/463

Comment (rhaamo) [reply ansible/roles/common/tasks/tools.yml:8]: https://unix.stackexchange.com/questions/625576/how-to-run-32-bit-armhf-binaries-on-64-bit-arm64-debian-os-on-raspberry-pi I think it could be possible (unless/until the CPU drops entirely 32bit), but I don't have an arm64 device with such userland to test

but if it's not possible at all then wxtoimg fate is uncertain on arm64 :(

**Comment by dom-robinson** [comment]
@Cadair in particular, but checking with @colinluthier too - looks ok to me, but can i get a second vote before i merge?

**Comment by colinluthier** [comment]
I think this needs to remain a draft till the documentation is caught up. Also, a couple of the issues he notes could be cleaned up first. I don't have time right now but I would like to see if settings.yml is getting checked for the new vars completely.

**Comment by rhaamo** [comment]
I can take a look at what can be updated in the docs this week.
For settings.yml IIRC I already updated the schema with the new keys.

File: ansible/roles/common/tasks/user.yml (hunk 3 above)
Comment (rhaamo) [reply ansible/roles/common/tasks/user.yml:1]: I haven't seen errors in logs, but I'm not really sure how this grc is called and when, my METEOR passes (gnuradio or rtl_fm) never generate anything here, probably too weak signal or something.

**Comment by colinluthier** [comment]
I just got a chance to read through it. settings.yml.sample was not changed and needs to be updated, unless I missed it. Also, the new variables need to be added to raspberry-noaa-v2/config/settings_schema.json. The PR looks great and we really appreciate the help.

**Comment by leonardpitzu** [comment]
No update on this one? I would like to migrate away from pi and more to docker stuff on an Intel based server. I'll keep the pi only as a remote sdr server connected to the antenna, in my garden, where it's hard for a server to sit.

**Comment by MihajloPi** [comment]
Implemented amd64 support alongside the native armhf for Raspberry Pi and X96 Android boxes with flashed Raspberry OS.
